File: Week 06/id_467/130被围绕的区域.cpp
class UnionFind {
public:
    vector<int> p;
    UnionFind (int size) {
        p.resize(size);
        for (int i = 0; i < size; ++i) {
            p[i] = i;
        }
    }

    void _union(int n1, int n2) {
        int p1 = find(n1);
        int p2 = find(n2);
        if (p1 != p2) {
            p[p2] = p1;
        }
    }

    int find(int t) {
        while (p[t] != t) {
            p[t] = p[p[t]];
            t = p[t];
        }
        return t;
    }

    bool isConnect(int n1, int n2) {
        return find(n1) == find(n2);
    }
};

class Solution {
public:
    int row;
    int col;
    void solve(vector<vector<char>>& board) {
        if (board.size() == 0) return;
        row = board.size();
        col = board[0].size();
        UnionFind uf = UnionFind(row * col + 1);
        int tmpnode = row * col;
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (board[i][j] == 'O') {
                    if (i == 0 || i == row - 1 || j == 0 || j == col - 1) {
                        uf._union(getIndex(i, j), tmpnode);
                    } else {
                        if (i > 0 && board[i-1][j] == 'O') {
                            uf._union(getIndex(i, j), getIndex(i-1, j));
                        }
                        if (i < row - 1 && board[i+1][j] == 'O') {
                            uf._union(getIndex(i, j), getIndex(i+1, j));
                        }
                        if (j > 0 && board[i][j-1] == 'O') {
                            uf._union(getIndex(i, j), getIndex(i, j-1));
                        }
                        if (j < col - 1 && board[i][j+1] == 'O') {
                            uf._union(getIndex(i, j), getIndex(i, j+1));
                        }                        
                    }
                }
            }
        }

        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (uf.isConnect(getIndex(i, j), tmpnode)) {
                    board[i][j] = 'O';
                } else {
                    board[i][j] = 'X';
                }
            }
        }
    }

    int getIndex(int i, int j) {
        return i * col + j;
    }
};

File: Week 06/id_477/130.被围绕的区域.go
package leetcode

/*
 * @lc app=leetcode.cn id=130 lang=golang
 *
 * [130] 被围绕的区域
 */

// @lc code=start
func solve(board [][]byte) {
	if len(board) == 0 || len(board[0]) == 0 {
		return
	}
	m, n := len(board), len(board[0])
	uf := unionFind(m*n + 1)
	dummyNode := m * n
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if board[i][j] == 'O' {
				if i == 0 || j == 0 || i == m-1 || j == n-1 {
					union(uf, i*n+j, dummyNode)
				} else {
					if i > 0 && board[i-1][j] == 'O' {
						union(uf, i*n+j, (i-1)*n+j)
					}
					if i < m-1 && board[i+1][j] == 'O' {
						union(uf, i*n+j, (i+1)*n+j)
					}
					if j > 0 && board[i][j-1] == 'O' {
						union(uf, i*n+j, i*n+j-1)
					}
					if j < n-1 && board[i][j+1] == 'O' {
						union(uf, i*n+j, i*n+j+1)
					}
				}
			}
		}
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if !isContact(uf, i*n+j, dummyNode) {
				board[i][j] = 'X'
			}
		}
	}
}

func unionFind(total int) []int {
	parents := make([]int, total)
	for i := range parents {
		parents[i] = i
	}
	return parents
}

func find(parent []int, i int) int {
	j := i
	for parent[i] != i {
		i = parent[i]
	}
	for parent[j] != j {
		j, parent[j] = parent[j], i
	}
	return i
}

func union(parent []int, i, j int) []int {
	p1 := find(parent, i)
	p2 := find(parent, j)
	if p1 != p2 {
		parent[p1] = p2
	}
	return parent
}

func isContact(parent []int, i, j int) bool {
	return find(parent, i) == find(parent, j)
}

// @lc code=end
